File: 2024/19/LinenLayout.py
with open("input") as f:
    raw = f.read()

towels, patterns = raw.split("\n\n")
towels = [s.strip() for s in towels.split(",")]
patterns = list(patterns.splitlines())


def solve(cur, path, memo, p2=False):
    if cur in memo:
        return memo[cur]
    if not cur:
        return 1
    total = 0
    for t in towels:
        if cur.startswith(t):
            memo[cur] = solve(cur.removeprefix(t), path + [t], memo, p2)
            if p2:
                total += memo[cur]
            else:
                total |= memo[cur]
    memo[cur] = total
    return total


print(sum(solve(p, [], {}) for p in patterns))
print(sum(solve(p, [], {}, True) for p in patterns))

File: 2024/19/LinenLayout.rs
use std::{
    collections::HashMap,
    ops::{Add, BitOr},
};

fn main() {
    let input = std::fs::read_to_string("input").unwrap();
    let (towels, patterns) = input.split_once("\n\n").unwrap();
    let mut towels: Vec<_> = towels.split(", ").collect();
    towels.sort_unstable_by_key(|n| n.len());
    let patterns = patterns.lines();
    let ans: u64 =
        patterns.clone().map(|p| solve(p, &towels, &mut HashMap::new(), u64::bitor)).sum();
    println!("{ans}");
    let ans: u64 = patterns.map(|p| solve(p, &towels, &mut HashMap::new(), u64::add)).sum();
    println!("{ans}");
}

fn solve<'a>(
    cur: &'a str,
    towels: &[&str],
    memo: &mut HashMap<&'a str, u64>,
    f: fn(u64, u64) -> u64,
) -> u64 {
    if cur.is_empty() {
        1
    } else if let Some(&n) = memo.get(cur) {
        n
    } else {
        towels
            .iter()
            .filter_map(|t| cur.strip_prefix(t))
            .map(|s| solve(s, towels, memo, f))
            .reduce(f)
            .inspect(|&n| _ = memo.insert(cur, n))
            .unwrap_or_default()
    }
}
